File: examples/CBuildr4/custord1.cpp
//---------------------------------------------------------------------------
#include <vcl\vcl.h>
#pragma hdrstop

#include <stdio.h>

#include "CustOrd1.h"
#include "CustOrd3.h"
#include "CustOrd2.h"
//---------------------------------------------------------------------------
#pragma link "OvcTable"
#pragma link "OvcTCmmn"
#pragma link "OvcBase"
#pragma link "OvcTCSim"
#pragma link "OvcTCBEF"
#pragma link "OvcTCStr"
#pragma link "OvcTCell"
#pragma link "OvcTCPic"
#pragma link "OvcTCHdr"
#pragma link "Ovccache"
#pragma resource "*.dfm"

TfrmCustomers *frmCustomers;

int MaxCacheItems = 20;

//---------------------------------------------------------------------------
__fastcall TfrmCustomers::TfrmCustomers(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::GetCustomerRecord(int RowNum, void* &P, Word& Size)
{
  // position db cursor to requested record
  PositionDatabase(RowNum);
  // return size of data
  Size = sizeof(TCustomerData);
  // allocate buffer for data
  P = new TCustomerData;
  // Create a TCustomerData reference to the void*. I used a reference
  // here but I could have used a pointer instead. Using a reference
  // allows us to use the dot operator rather than the -> operator.
  // To see the difference look at the corresponding method in the
  // CUSTORD2.CPP file. There I used a pointer rather than a reference.
  TCustomerData& data = *reinterpret_cast<TCustomerData*>(P);
  // read the db record
  data.CustNo  = dbCustomerCustNo->AsFloat;
  data.Company = dbCustomerCompany->AsString;
  data.Addr1 =dbCustomerAddr1->AsString;
  data.Addr2 =dbCustomerAddr2->AsString;
  data.City =dbCustomerCity->AsString;
  data.State =dbCustomerState->AsString;
  data.Zip =dbCustomerZip->AsString;
  data.Country =dbCustomerCountry->AsString;
  data.Phone =dbCustomerPhone->AsString;
  data.FAX =dbCustomerFAX->AsString;
  data.TaxRate = dbCustomerTaxRate->AsFloat;
  data.Contact =dbCustomerContact->AsString;
  data.InvoiceDate = DateTimeToStDate(dbCustomerLastInvoiceDate->AsDateTime);
}

void TfrmCustomers::PositionDatabase(long RowNum)
{
  static long LastRecordRequest = -1;
  // position db cursor to requested record
  if (LastRecordRequest == -1) {
    dbCustomer->First();
    dbCustomer->MoveBy(RowNum - 1);
  }
  else if (RowNum == 1)
    dbCustomer->First();
  else if (RowNum == dbCustomer->RecordCount)
    dbCustomer->Last();
  else
    dbCustomer->MoveBy(RowNum - LastRecordRequest);
  LastRecordRequest = RowNum;
}

void TfrmCustomers::RefreshRecord(long RecNum)
{
  CustomerCache->Update(RecNum);
  tblCustomer->AllowRedraw = false;
  tblCustomer->InvalidateRow(RecNum);
  tblCustomer->AllowRedraw = true;
}


void __fastcall TfrmCustomers::tblCustomerGetCellData(TObject *Sender,
  int RowNum, int ColNum, Pointer &Data, TOvcCellDataPurpose Purpose)
{
  TCustomerData* P = 0;
  CellData = "?";
  Data = &CellData;
  if (RowNum == 0) {
    switch (ColNum) {
      case 0 : CellData = ""; break;
      case 1 : CellData = dbCustomerCustNo->DisplayLabel; break;
      case 2 : CellData = dbCustomerCompany->DisplayLabel; break;
      case 3 : CellData = dbCustomerAddr1->DisplayLabel; break;
      case 4 : CellData = dbCustomerAddr2->DisplayLabel; break;
      case 5 : CellData = dbCustomerCity->DisplayLabel; break;
      case 6 : CellData = dbCustomerState->DisplayLabel; break;
      case 7 : CellData = dbCustomerZip->DisplayLabel; break;
      case 8 : CellData = dbCustomerCountry->DisplayLabel; break;
      case 9 : CellData = dbCustomerPhone->DisplayLabel; break;
      case 10 : CellData = dbCustomerFAX->DisplayLabel; break;
      case 11 : CellData = dbCustomerTaxRate->DisplayLabel; break;
      case 12 : CellData = dbCustomerContact->DisplayLabel; break;
      case 13 : CellData = dbCustomerLastInvoiceDate->DisplayLabel;
    }
  }
  else {
    P = reinterpret_cast<TCustomerData*>(CustomerCache->Items[RowNum]);
    if (P) {
      switch (ColNum) {
        case 0 :
        case 1 : Data = &P->CustNo; break;
        case 2 : Data = &P->Company; break;
        case 3 : Data = &P->Addr1; break;
        case 4 : Data = &P->Addr2; break;
        case 5 : Data = &P->City; break;
        case 6 : Data = &P->State; break;
        case 7 : Data = &P->Zip; break;
        case 8 : Data = &P->Country; break;
        case 9 : Data = &P->Phone; break;
        case 10 : Data = &P->FAX; break;
        case 11 : Data = &P->TaxRate; break;
        case 12 : Data = &P->Contact; break;
        case 13 : Data = &P->InvoiceDate;
      }
    }
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::FormCreate(TObject *Sender)
{
  CustomerCache = new TOvcCache(this);
  CustomerCache->MaxCacheItems = MaxCacheItems;
  CustomerCache->OnGetItem = &GetCustomerRecord;

  dbCustomer->Active = true;
  if (dbCustomer->Active)
    tblCustomer->RowLimit = dbCustomer->RecordCount + 1;
  else
    tblCustomer->RowLimit = 1;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::FormDestroy(TObject *Sender)
{
  dbCustomer->Active = false;
  delete CustomerCache;
  CustomerCache = 0;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::btnOrdersClick(TObject *Sender)
{
  double CustNo;
  // if editing, stop and post changes
  if (tblCustomer->InEditingState())
    if (!tblCustomer->StopEditingState(true))
      return;

  frmOrders = new TfrmOrders(this);
  Screen->Cursor = crHourGlass;
  try {
    CustNo = reinterpret_cast<TCustomerData*>
      ((CustomerCache->Items[tblCustomer->ActiveRow]))->CustNo;
    frmOrders->UpdateOrderTable(CustNo);
    char buff[40];
    sprintf(buff, "Orders for customer# %.0f (Read-Only)", CustNo);
    frmOrders->Panel2->Caption = buff;
    frmOrders->ShowModal();
  }
  catch (...) {
    Screen->Cursor = crDefault;
    delete frmOrders;
    throw;
  }
  Screen->Cursor = crDefault;
  delete frmOrders;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::tblCustomerBeginEdit(TObject *Sender,
  int RowNum, int ColNum, bool &AllowIt)
{
  // refresh current record
  RefreshRecord(RowNum);

  // can this record be edited (Col 1 is the customer number)
  AllowIt = ((ColNum > 1) && dbCustomer->CanModify);
  if (AllowIt) {
    try {
      dbCustomer->Edit();
    }
    catch (...) {
      AllowIt = false;
      throw;
    }
    // lock record so that it stays in the cache until we are done editing
    CustomerCache->LockCacheItem(RowNum);
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::tblCustomerDoneEdit(TObject *Sender, int RowNum,
  int ColNum)
{
  TCustomerData* P = 0;
  try {
    // update database record
    P  = reinterpret_cast<TCustomerData*>(CustomerCache->Items[RowNum]);
    if (P) {
      switch (ColNum) {
        case 0 : // row label column
        case 1 : // no editing of customer number allowed
        case 2 : dbCustomerCompany->AsString = P->Company; break;
        case 3 : dbCustomerAddr1->AsString = P->Addr1; break;
        case 4 : dbCustomerAddr2->AsString = P->Addr2; break;
        case 5 : dbCustomerCity->AsString = P->City; break;
        case 6 : dbCustomerState->AsString = P->State; break;
        case 7 : dbCustomerZip->AsString = P->Zip; break;
        case 8 : dbCustomerCountry->AsString = P->Country; break;
        case 9 : dbCustomerPhone->AsString = P->Phone; break;
        case 10 : dbCustomerFAX->AsString = P->FAX; break;
        case 11 : dbCustomerTaxRate->AsFloat = P->TaxRate; break;
        case 12 : dbCustomerContact->AsString = P->Contact; break;
        case 13 : dbCustomerLastInvoiceDate->AsDateTime =
              StDateToDateTime(P->InvoiceDate);
      }
      // post changes and stop editing
      dbCustomer->Post();
    }
  }
  catch (...) {
    CustomerCache->UnlockCacheItem();
    throw;
  }
  // unlock record
  CustomerCache->UnlockCacheItem();
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::btnCloseClick(TObject *Sender)
{
  // if editing, stop and post changes
  if (tblCustomer->InEditingState())
    if (!tblCustomer->StopEditingState(true))
      return;
  Close();
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::tblCustomerActiveCellChanged(TObject *Sender,
  int RowNum, int ColNum)
{
  static int LastRow = -1;
  // refresh row
  if (LastRow != RowNum) {
    LastRow = RowNum;
    RefreshRecord(RowNum);
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::tblCustomerTopLeftCellChanging(TObject *Sender,
  int &RowNum, int &ColNum)
{
  int R, C;
  R = tblCustomer->TopRow;
  C = tblCustomer->LeftCol;
  if (RowNum == R) return;

  try {
    // if editing, stop and post changes
    if (tblCustomer->InEditingState())
      if (!tblCustomer->StopEditingState(true)) {
        RowNum = R;
        ColNum = C;
      }
  }
  catch (...) {
    RowNum = R;
    ColNum = C;
    throw;
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::tblCustomerLockedCellClick(TObject *Sender,
  int RowNum, int ColNum)
{
  if (RowNum == 0)
    tblCustomer->ActiveCol = ColNum;
  if (ColNum == 0)
    tblCustomer->ActiveRow = RowNum;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCustomers::btnAboutClick(TObject *Sender)
{
  frmAbout = new TfrmAbout(this);
  frmAbout->ShowModal();
  delete frmAbout;
}
//---------------------------------------------------------------------------

File: examples/Delphi/CustOrd1.pas
{$J+} {Writable constants}

unit CustOrd1;

interface

uses
  SysUtils, WinTypes, WinProcs, Messages, Classes, Graphics, Controls,
  Forms, DBTables, DB, OvcBase, OvcTCmmn, OvcTable, OvcTCStr, OvcTCEdt,

  OvcCache, {This is a bonus class contained in the Orpheus\bonus
  directory, it should be manually added to the project}

  OvcData, OvcEF, OvcSF, OvcTCSim, OvcTCBEF,
  OvcPB, OvcPF, OvcTCPic, StdCtrls, Buttons, CustOrd2, CustOrd3,
  ExtCtrls, OvcTCHdr, OvcTCell, OvcDate;

const
  {number of records to cache}
  MaxCacheItems = 20;

type
  {definition of one record for the customer table}
  PCustomerData = ^TCustomerData;
  TCustomerData = record
    CustNo      : Double; {*index}
    Company     : string[30];
    Addr1       : string[30];
    Addr2       : string[30];
    City        : string[15];
    State       : string[20];
    Zip         : string[10];
    Country     : string[20];
    Phone       : string[15];
    FAX         : string[15];
    TaxRate     : Double;
    Contact     : string[20];
    InvoiceDate : TOvcDate;
  end;

type
  TfrmCustomers = class(TForm)
    OvcController1: TOvcController;
    tblCustomer: TOvcTable;
    dbCustomer: TTable;
    dbCustomerCustNo: TFloatField;
    dbCustomerCompany: TStringField;
    dbCustomerAddr1: TStringField;
    dbCustomerAddr2: TStringField;
    dbCustomerCity: TStringField;
    dbCustomerState: TStringField;
    dbCustomerZip: TStringField;
    dbCustomerCountry: TStringField;
    dbCustomerPhone: TStringField;
    dbCustomerFAX: TStringField;
    dbCustomerTaxRate: TFloatField;
    dbCustomerContact: TStringField;
    dbCustomerLastInvoiceDate: TDateTimeField;
    sfCustNo1: TOvcTCSimpleField;
    sfCompany: TOvcTCSimpleField;
    sfAddr1: TOvcTCSimpleField;
    sfAddr2: TOvcTCSimpleField;
    sfCity: TOvcTCSimpleField;
    sfState: TOvcTCSimpleField;
    pfZip: TOvcTCPictureField;
    sfCountry: TOvcTCSimpleField;
    pfPhone: TOvcTCPictureField;
    pfFax: TOvcTCPictureField;
    pfTaxRate1: TOvcTCPictureField;
    sfContact: TOvcTCSimpleField;
    pfLastInvoiceDate: TOvcTCPictureField;
    Panel1: TPanel;
    RowLabel: TOvcTCRowHead;
    btnOrders: TBitBtn;
    btnClose: TBitBtn;
    dbCustomerHeaderCell: TOvcTCColHead;
    btnAbout: TBitBtn;
    procedure tblCustomerGetCellData(Sender: TObject; RowNum: Longint;
      ColNum: Integer; var Data: Pointer; Purpose : TOvcCellDataPurpose);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure btnOrdersClick(Sender: TObject);
    procedure tblCustomerBeginEdit(Sender: TObject; RowNum: Longint;
      ColNum: Integer; var AllowIt: Boolean);
    procedure tblCustomerDoneEdit(Sender: TObject; RowNum: Longint;
      ColNum: Integer);
    procedure btnCloseClick(Sender: TObject);
    procedure tblCustomerActiveCellChanged(Sender: TObject;
      RowNum: Longint; ColNum: Integer);
    procedure tblCustomerTopLeftCellChanging(Sender: TObject;
      var RowNum: Longint; var ColNum: Integer);
    procedure tblCustomerLockedCellClick(Sender: TObject; RowNum: Longint;
      ColNum: Integer);
    procedure btnAboutClick(Sender: TObject);

  private
    CellData      : string;
    CustomerCache : TOvcCache;

  public
    procedure GetCustomerRecord(RowNum : LongInt; var P : Pointer; var Size : Word);
    procedure PositionDatabase(RowNum : LongInt);
    procedure RefreshRecord(RecNum : LongInt);
  end;

var
  frmCustomers: TfrmCustomers;

implementation

{$R *.DFM}

procedure TfrmCustomers.PositionDatabase(RowNum : LongInt);
const
  LastRecordRequest : LongInt = -1;
begin
  {position db cursor to requested record}
  if LastRecordRequest = -1 then begin
    dbCustomer.First;
    dbCustomer.MoveBy(RowNum-1);
  end else if RowNum = 1 then
    dbCustomer.First
  else if RowNum = dbCustomer.RecordCount then
    dbCustomer.Last
  else
    dbCustomer.MoveBy(RowNum-LastRecordRequest);
  LastRecordRequest := RowNum;
end;

procedure TfrmCustomers.GetCustomerRecord(RowNum : LongInt; var P : Pointer; var Size : Word);
begin
  {position db cursor to requested record}
  PositionDatabase(RowNum);
  {return size of data}
  Size := SizeOf(TCustomerData);
  {allocate buffer for data}
  GetMem(P, Size);
  {read the db record}
  with PCustomerData(P)^ do begin
    CustNo      := dbCustomerCustNo.AsFloat;
    Company     := ShortString(dbCustomerCompany.AsString);
    Addr1       := ShortString(dbCustomerAddr1.AsString);
    Addr2       := ShortString(dbCustomerAddr2.AsString);
    City        := ShortString(dbCustomerCity.AsString);
    State       := ShortString(dbCustomerState.AsString);
    Zip         := ShortString(dbCustomerZip.AsString);
    Country     := ShortString(dbCustomerCountry.AsString);
    Phone       := ShortString(dbCustomerPhone.AsString);
    FAX         := ShortString(dbCustomerFAX.AsString);
    TaxRate     := dbCustomerTaxRate.AsFloat;
    Contact     := ShortString(dbCustomerContact.AsString);
    InvoiceDate := DateTimeToStDate(dbCustomerLastInvoiceDate.AsDateTime);
  end;
end;

procedure TfrmCustomers.tblCustomerGetCellData(Sender: TObject; RowNum: Longint;
  ColNum: Integer; var Data: Pointer; Purpose : TOvcCellDataPurpose);
var
  P : PCustomerData;
begin
  CellData := '?';
  Data := @CellData;
  if RowNum = 0 then begin
    case ColNum of
      0 : CellData := '';
      1 : CellData := dbCustomerCustNo.DisplayLabel;
      2 : CellData := dbCustomerCompany.DisplayLabel;
      3 : CellData := dbCustomerAddr1.DisplayLabel;
      4 : CellData := dbCustomerAddr2.DisplayLabel;
      5 : CellData := dbCustomerCity.DisplayLabel;
      6 : CellData := dbCustomerState.DisplayLabel;
      7 : CellData := dbCustomerZip.DisplayLabel;
      8 : CellData := dbCustomerCountry.DisplayLabel;
      9 : CellData := dbCustomerPhone.DisplayLabel;
     10 : CellData := dbCustomerFAX.DisplayLabel;
     11 : CellData := dbCustomerTaxRate.DisplayLabel;
     12 : CellData := dbCustomerContact.DisplayLabel;
     13 : CellData := dbCustomerLastInvoiceDate.DisplayLabel;
    end;
  end else begin
    P  := CustomerCache.Items[RowNum];
    if Assigned(P) then begin
      case ColNum of
        0 : {};
        1 : Data := @P^.CustNo;
        2 : Data := @P^.Company;
        3 : Data := @P^.Addr1;
        4 : Data := @P^.Addr2;
        5 : Data := @P^.City;
        6 : Data := @P^.State;
        7 : Data := @P^.Zip;
        8 : Data := @P^.Country;
        9 : Data := @P^.Phone;
       10 : Data := @P^.FAX;
       11 : Data := @P^.TaxRate;
       12 : Data := @P^.Contact;
       13 : Data := @P^.InvoiceDate;
      end
    end;
  end;
end;

procedure TfrmCustomers.FormCreate(Sender: TObject);
begin
  CustomerCache := TOvcCache.Create(Self);
  CustomerCache.MaxCacheItems := MaxCacheItems;
  CustomerCache.OnGetItem := GetCustomerRecord;

  dbCustomer.Active := True;
  if dbCustomer.Active then
    tblCustomer.RowLimit := dbCustomer.RecordCount + 1
  else
    tblCustomer.RowLimit := 1;
end;

procedure TfrmCustomers.FormDestroy(Sender: TObject);
begin
  dbCustomer.Active := False;
  CustomerCache.Free;
  CustomerCache := nil;
end;

procedure TfrmCustomers.btnOrdersClick(Sender: TObject);
var
  CustNo : Double;
begin
  {if editing, stop and post changes}
  if tblCustomer.InEditingState then
    if not tblCustomer.StopEditingState(True) then
      Exit;

  frmOrders := TFrmOrders.Create(Self);
  try
    Screen.Cursor := crHourGlass;
    try
      CustNo := PCustomerData(CustomerCache.Items[tblCustomer.ActiveRow])^.CustNo;
      frmOrders.UpdateOrderTable(CustNo);
    finally
      Screen.Cursor := crDefault;
    end;
    frmOrders.Panel2.Caption :=
      Format('Orders for customer# %.0f (Read-Only)', [CustNo]);
    frmOrders.ShowModal;
  finally
    frmOrders.Free;
  end;
end;

procedure TfrmCustomers.RefreshRecord(RecNum : LongInt);
begin
  if Assigned(CustomerCache) then begin                                {!!.13}
    CustomerCache.Update(RecNum);
    tblCustomer.AllowRedraw := False;
    tblCustomer.InvalidateRow(RecNum);
    tblCustomer.AllowRedraw := True;
  end;                                                                 {!!.13}
end;

procedure TfrmCustomers.tblCustomerBeginEdit(Sender: TObject;
  RowNum: Longint; ColNum: Integer; var AllowIt: Boolean);
begin
  {refresh current record}
  RefreshRecord(RowNum);

  {can this record be edited (Col 1 is the customer number)}
  AllowIt := (ColNum > 1) and dbCustomer.CanModify;
  if AllowIt then begin
    try
      dbCustomer.Edit;
    except
      AllowIt := False;
      raise
    end;
    {lock record so that it stays in the cache until we are done editing}
    CustomerCache.LockCacheItem(RowNum);
  end;
end;

procedure TfrmCustomers.tblCustomerDoneEdit(Sender: TObject;
  RowNum: Longint; ColNum: Integer);
var
  P : PCustomerData;
begin
  try
    {update database record}
    P  := PCustomerData(CustomerCache.Items[RowNum]);
    if Assigned(P) then begin
      case ColNum of
        0 : {row label column};
        1 : {no editing of customer number allowed};
        2 : dbCustomerCompany.AsString := string(P^.Company);
        3 : dbCustomerAddr1.AsString   := string(P^.Addr1);
        4 : dbCustomerAddr2.AsString   := string(P^.Addr2);
        5 : dbCustomerCity.AsString    := string(P^.City);
        6 : dbCustomerState.AsString   := string(P^.State);
        7 : dbCustomerZip.AsString     := string(P^.Zip);
        8 : dbCustomerCountry.AsString := string(P^.Country);
        9 : dbCustomerPhone.AsString   := string(P^.Phone);
       10 : dbCustomerFAX.AsString     := string(P^.FAX);
       11 : dbCustomerTaxRate.AsFloat  := P^.TaxRate;
       12 : dbCustomerContact.AsString := string(P^.Contact);
       13 : dbCustomerLastInvoiceDate.AsDateTime :=
              StDateToDateTime(P^.InvoiceDate);
      end;
      {post changes and stop editing}
      dbCustomer.Post;
    end;
  finally
    {unlock record}
    CustomerCache.UnlockCacheItem;
  end;
end;

procedure TfrmCustomers.btnCloseClick(Sender: TObject);
begin
  {if editing, stop and post changes}
  if tblCustomer.InEditingState then
    if not tblCustomer.StopEditingState(True) then
      Exit;
  Close;
end;

procedure TfrmCustomers.tblCustomerActiveCellChanged(Sender: TObject;
  RowNum: Longint; ColNum: Integer);
const
  LastRow : LongInt = -1;
begin
  {refresh row}
  if LastRow <> RowNum then begin
    LastRow := RowNum;
    RefreshRecord(RowNum);
  end;
end;

procedure TfrmCustomers.tblCustomerTopLeftCellChanging(Sender: TObject;
  var RowNum: Longint; var ColNum: Integer);
var
  R : LongInt;
  C : Integer;
begin
  R := tblCustomer.TopRow;
  C := tblCustomer.LeftCol;
  if RowNum = R then Exit;

  try
    {if editing, stop and post changes}
    if tblCustomer.InEditingState then
      if not tblCustomer.StopEditingState(True) then begin
        RowNum := R;
        ColNum := C;
      end;
  except
    RowNum := R;
    ColNum := C;
    raise;
  end;
end;

procedure TfrmCustomers.tblCustomerLockedCellClick(Sender: TObject;
  RowNum: Longint; ColNum: Integer);
begin
  if RowNum = 0 then
    tblCustomer.ActiveCol := ColNum;
  if ColNum = 0 then
    tblCustomer.ActiveRow := RowNum;
end;

procedure TfrmCustomers.btnAboutClick(Sender: TObject);
begin
  frmAbout := TfrmAbout.Create(Self);
  try
    frmAbout.ShowModal;
  finally
    frmAbout.Free;
  end;
end;

end.
